### PR TITLE
Fixed some issues of the AccessDecisionManager profiler

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q | A |
| --- | --- |
| Branch? | 3.1 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #19022 https://github.com/symfony/symfony-standard/issues/968 https://github.com/schmittjoh/JMSSecurityExtraBundle/issues/207 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php [hunk 1]
```diff
@@ -66,7 +66,7 @@ public function load(array $configs, ContainerBuilder $container)
         $loader->load('collectors.xml');
         $loader->load('guard.xml');
 
-        if ($container->hasParameter('kernel.debug') && $container->getParameter('kernel.debug')) {
+        if ($container->has('profiler')) {
```
Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: this is broken, as it will always be false here (DI extensions are isolated from each other). You may need to use a compiler pass instead

File: src/Symfony/Component/Security/Core/Authorization/DebugAccessDecisionManager.php [hunk 2]
```diff
@@ -29,7 +29,7 @@ class DebugAccessDecisionManager implements AccessDecisionManagerInterface
     private $voters;
     private $decisionLog = array();
 
-    public function __construct(AccessDecisionManager $manager)
+    public function __construct(AccessDecisionManagerInterface $manager)
```
Comment (stof) [inline src/Symfony/Component/Security/Core/Authorization/DebugAccessDecisionManager.php]: the retrieval of the strategy with reflection below should check for being an instance of AccessDecisionManager though (btw, the ReflectionProperty should use ``AccessDecisionManager::class`` as first argument to be compatible with inheritance, as this is a private property which is not accessible on a child class)

Comment (stof) [inline src/Symfony/Component/Security/Core/Authorization/DebugAccessDecisionManager.php]: and ``setVoters`` should check that the wrapped implementation has this method too

Comment (Ener-Getick) [inline src/Symfony/Component/Security/Core/Authorization/DebugAccessDecisionManager.php]: :+1: I missed this PR, so in case it helps: https://github.com/symfony/symfony/compare/3.1...Ener-Getick:SECURITY

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php (hunk 1 above)
Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: This looks overkill to me. Isn't there any other solution? In other extensions, for example in [these lines](https://github.com/symfony/symfony/blob/2.3/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L371-378) of `FrameworkExtension`, they use the same code as here:

```php
```

Comment (WouterJ) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: The big difference is that parameters are copied to the local container passed to  `Extension#load()`, whereas services aren't.

Comment (WouterJ) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: It's about this:
```php
if ($container->has('profiler')) {
    // ...
}
```
Versus this:
```php
if ($container->getParameter('kernel.debug')) {
    // ...
}
```

The `$container` parameter of `Extension::load()` is *not* the actual container, it's a copy. See https://github.com/symfony/dependency-injection/blob/master/Compiler/MergeExtensionConfigurationPass.php#L47-L49 Only the parameters are copied from actual to tmp container, services aren't.

This is why `$container->has('profiler')` will *always* return false.

Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: Sure. I didn't use `$container->has('profiler')` at first. I used `$container->getParameter('kernel.debug')` but I was told that it was not completely right. But I still don't understand, so I can't fix it.

Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: @javiereguiluz the issue is that kernel.debug is only the default value for enabling the profiler. It is possible to enable it in non-debug mode.

And this mismatch is precisely one of the issues you have to fix, as the profiler always assumes that the debug access decision manager is there.
The alternative would be to handle properly the absence of this service (storing less info in the profiler panel)

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php [hunk 3]
```diff
@@ -66,7 +66,7 @@ public function load(array $configs, ContainerBuilder $container)
         $loader->load('collectors.xml');
         $loader->load('guard.xml');
 
-        if ($container->hasParameter('kernel.debug') && $container->getParameter('kernel.debug')) {
+        if ($container->getParameter('kernel.debug')) {
             $loader->load('security_debug.xml');
```
Comment (yceruto) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: @javiereguiluz I tried this PR and the issue still exists :(

https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/SecurityBundle/Resources/config/collectors.xml#L13: The `data_collector.security` service ("which is always loaded") require `debug.security.access.decision_manager`, so when `$container->getParameter('kernel.debug')` is `false` then:
```
  [Symfony\Component\DependencyInjection\Exception\ServiceNotFoundException]
  The service "profiler" has a dependency on a non-existent service "debug.security.access.decision_manager".
```

The road should be here https://github.com/symfony/symfony-standard/issues/968#issuecomment-223022665:
> ... It should be based on whether the profiler is enabled instead of being based on the debug mode ...

Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: Is it OK to load the `collectors.xml` file unconditionally? Is it really used when `debug = false` ? If this can be changed, the solution would be trivial: move `$loader->load('collectors.xml');` under the `if ($container->getParameter('kernel.debug'))` condition.

Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: well, if you do it, the security panel would not appear when enabling the profiler in no-debug mode, which might be confusing

Comment (yceruto) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: Could we check [`web_profiler.controller.profiler`](https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/WebProfilerBundle/Resources/config/profiler.xml#L8) service definition ? it's defined only if WebProfileBundle is enabled. BUT, this would imply that all panels profile should be loaded only if this "WebProfileBundle" is enabled.

Comment (stof) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: @yceruto WebProfilerBundle is not related to enabling the profiler (the bundle does not provide the profiler. It is only about providing a visualization UI).
The proper service would be detecting ``profiler`` (coming from FrameworkBundle), but this cannot be done in a DI extension, because DI extension cannot see services defined in other extensions (see the previous discussion)

Comment (yceruto) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: @stof you're right, then

https://github.com/symfony/symfony/pull/18934#discussion_r65501397
> ... Only the parameters are copied from actual to tmp container ...

We can create a new parameter something like `profiler.collect` into [`FrameworkExtension::registerProfilerConfiguration`](https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/FrameworkExtension.php#L291) ?:
```php
$container->setParameter('profiler.collect', $config['collect']);
```
and use it here ?
```php
if ($container->hasParameter('profiler.collect') && $container->getParameter('profiler.collect')) {
    $loader->load('collectors.xml');
    $loader->load('security_debug.xml');
}
```
@javiereguiluz WDYT?

**Comment by fabpot** [comment]
@javiereguiluz Can you move this PR forward? Or do you need some help?

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php (hunk 3 above)
Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: @yceruto my guess is that they won't accept to introduce a parameter for this. Let's try to do the compilar pass that @stof suggested in the first place.

**Comment by peterrehm** [comment]
@javiereguiluz Any update on this? A solution would be awesome.

**Comment by javiereguiluz** [comment]
@peterrehm I plan to finish this very soon.

**Comment by javiereguiluz** [comment]
I've updated this with a compiler pass, as @stof suggested.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php [hunk 4]
```diff
@@ -66,10 +66,6 @@ public function load(array $configs, ContainerBuilder $container)
         $loader->load('collectors.xml');
```
Comment (yceruto) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: @javiereguiluz the issue still occurs if `profiler` service is not defined, so `data_collector.security` require of `debug.security.access.decision_manager`

Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: Could we solve this via config?

```xml
<argument type="service" id="debug.security.access.decision_manager" on-invalid="ignore" />
```

Comment (HeahDude) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: Or maybe just load the collector in debug?

Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: @HeahDude apparently we can't do that. See https://github.com/symfony/symfony/pull/18934#discussion_r65502752

Comment (HeahDude) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: Ok, I see another solution, now that the `security.access.decision_manager` is properly decorated, just pass it to the collector instead of the `debug.security.access.decision_manager`, [this line](https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/SecurityBundle/DataCollector/SecurityDataCollector.php#L114) should ensure everything is fine, right?

Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: 👍 Let's try that! Thanks.

**Comment by javiereguiluz** [comment]
Some tests are still failing. Given that a new Symfony 3.1 release is imminent, could someone please help me fix this? Thanks!

**Comment by HeahDude** [comment]
If passing the `security.access.decision_manager` instead of the `debug.security.access.decision_manager` to the collector maybe you can revert the compiler pass and keep the xml config.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php [hunk 5]
```diff
@@ -65,10 +65,7 @@ public function load(array $configs, ContainerBuilder $container)
         $loader->load('templating_twig.xml');
         $loader->load('collectors.xml');
         $loader->load('guard.xml');
-
-        if ($container->hasParameter('kernel.debug') && $container->getParameter('kernel.debug')) {
-            $loader->load('security_debug.xml');
-        }
+        $loader->load('security_debug.xml');
```
Comment (HeahDude) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: Why did you remove the debug check?

File: src/Symfony/Bundle/SecurityBundle/SecurityBundle.php [hunk 6]
```diff
@@ -13,6 +13,7 @@
 
 use Symfony\Component\HttpKernel\Bundle\Bundle;
 use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Bundle\SecurityBundle\DependencyInjection\Compiler\AddDebugAccessDecisionManagerPass;
```
Comment (HeahDude) [inline src/Symfony/Bundle/SecurityBundle/SecurityBundle.php]: Can be removed :)

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php (hunk 5 above)
Comment (javiereguiluz) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: https://github.com/symfony/symfony/pull/18934#discussion_r65502752

Comment (HeahDude) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: Yes but now the check is done in the collector, so don't need to load the debug add when not in debug mode.

File: src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php [hunk 7]
```diff
@@ -66,7 +66,7 @@ public function load(array $configs, ContainerBuilder $container)
         $loader->load('collectors.xml');
         $loader->load('guard.xml');
 
-        if ($container->hasParameter('kernel.debug') && $container->getParameter('kernel.debug')) {
+        if ($container->getParameter('kernel.debug')) {
```
Comment (HeahDude) [inline src/Symfony/Bundle/SecurityBundle/DependencyInjection/SecurityExtension.php]: You should keep the check with `hasParameter()` since some tests are still failing

**Comment by HeahDude** [comment]
I guess you now need to update the `SecurityBundle` `composer.json` to require `"symfony/security": "~3.1,>=3.1.2"`

**Comment by HeahDude** [comment]
👍 LGTM now, remaining failures are unrelated.

Status: Reviewed

File: src/Symfony/Bundle/SecurityBundle/Resources/config/collectors.xml [hunk 8]
```diff
@@ -10,7 +10,7 @@
             <argument type="service" id="security.token_storage" on-invalid="ignore" />
             <argument type="service" id="security.role_hierarchy" />
             <argument type="service" id="security.logout_url_generator" />
-            <argument type="service" id="debug.security.access.decision_manager" />
+            <argument type="service" id="security.access.decision_manager" />
```
Comment (yceruto) [inline src/Symfony/Bundle/SecurityBundle/Resources/config/collectors.xml]: if the compiler pass was removed then when this argument is changed by `debug.security.access.decision_manager` ? I missing something ?

Comment (HeahDude) [inline src/Symfony/Bundle/SecurityBundle/Resources/config/collectors.xml]: @yceruto the service is decorated the debug is on by loading `security_debug.xml`.

**Comment by fabpot** [comment]
Thank you @javiereguiluz.

**Comment by loostro** [comment]
It's in the 3.1.x-dev branch, but not in the 3.1.1 tag :( @fabpot could you tag another patch release?

**Comment by peterrehm** [comment]
@loostro I think you need to be a bit patient, as @javiereguiluz wrote a new release is around the corner...
